### PR TITLE
[plugin-mobile-app] Add logic for verifying element visibility after swipe

### PR DESCRIPTION
**Pull request by VolhaHurynovich**

### CODE REVIEW

**Review by valfirst — COMMENTED** [review]
http://chris.beams.io/posts/git-commit/

File: vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/steps/TouchSteps.java [hunk 1]
```diff
@@ -134,27 +134,50 @@ private Supplier<Rectangle> createSwipeArea()
...
-        int visibilityY = swipeAreaSizeHeight - bottomVisibilityIndent;
-        if (elementTopCoordinateY > visibilityY)
+        int swipeAreaSize;
+        int elementCoordinate;
+        int bottomOrLeftVisibilityIndent = 0;
+        int topOrRightVisibilityIndent = 0;
+        if (SwipeDirection.LEFT == direction || SwipeDirection.RIGHT == direction)
+        {
+            swipeAreaSize = swipeArea.getWidth();
+            elementCoordinate = SwipeDirection.LEFT == direction
```
Comment (valfirst) [inline vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/steps/TouchSteps.java:147]: avoid duplicate checks

File: vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/action/TouchActions.java [hunk 2]
```diff
@@ -224,10 +224,24 @@ private Supplier<BufferedImage> createScreenShooter(SearchContext searchContext)
...
+     * @param startX        start X coordinate
+     * @param endX          end X coordinate
+     * @param swipeArea     the area to execute the swipe
+     * @param swipeDuration swipe duration in <a href="https://en.wikipedia.org/wiki/ISO_8601">ISO 8601</a> format
+     */
+    public void performHorizontalSwipe(int startX, int endX, Rectangle swipeArea, Duration swipeDuration)
+    {
+        swipe(SwipeDirection.createXCoordinates(startX, endX, swipeArea.getHeight(),
+                mobileApplicationConfiguration.getSwipeHorizontalYPosition(), swipeArea.getPoint()), swipeDuration);
+    }
```
Comment (valfirst) [inline vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/action/TouchActions.java:243]: why not to have a single method, that will accept swipe direction as an input parameter?

File: vividus-plugin-mobile-app/src/test/java/org/vividus/mobileapp/steps/TouchStepsTests.java [hunk 3]
```diff
@@ -130,14 +132,28 @@ void shouldSwipeToElement(int elementY, int endY)
...
+        {
+            when(element.getSize()).thenReturn(new Dimension(10, 10));
+        }
+        touchSteps.swipeToElement(direction, locator, Duration.ZERO);
 
         verifyNoMoreInteractions(parameters);
-        verify(touchActions).performVerticalSwipe(eq(592), eq(endY), argThat(sa ->
-            DIMENSION.equals(sa.getDimension()) && sa.getPoint().equals(new Point(0, 0))), eq(Duration.ZERO));
+
+        if (SwipeDirection.LEFT == direction || SwipeDirection.RIGHT == direction)
```
Comment (valfirst) [inline vividus-plugin-mobile-app/src/test/java/org/vividus/mobileapp/steps/TouchStepsTests.java:145]: unit tests shouldn't contain business logic

File: vividus-plugin-mobile-app/src/test/java/org/vividus/mobileapp/steps/TouchStepsTests.java [hunk 4]
```diff
@@ -197,35 +213,24 @@ void shouldSwipeToElementToTryToFindItButThatDoesntExist()
...
     @ParameterizedTest
-    @ValueSource(strings = {"UP", "DOWN"})
+    @ValueSource(strings = {"UP", "DOWN", "LEFT", "RIGHT"})
     void shouldNotSwipeToElementIfItAlreadyExists(SwipeDirection swipeDireciton)
     {
         mockScreenSize();
         WebElement element = mock(WebElement.class);
-        when(element.getLocation()).thenReturn(new Point(-1, 500));
+        when(element.getLocation()).thenReturn(new Point(20, 500));
+        if (SwipeDirection.LEFT == swipeDireciton)
```
Comment (valfirst) [inline vividus-plugin-mobile-app/src/test/java/org/vividus/mobileapp/steps/TouchStepsTests.java:222]: unit tests shouldn't contain business logic

**Review by valfirst — CHANGES_REQUESTED** [review]

File: vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/model/SwipeDirection.java [hunk 5]
```diff
@@ -78,13 +75,20 @@ public SwipeCoordinates calculateCoordinates(Rectangle swipeArea,
...
     {
         int x = calculateCoordinate(width, xOffsetPercentage);
         return createAdjustedCoordinates(x, startY, x, endY, point);
     }
 
+    public static SwipeCoordinates createXCoordinates(int startX, int endX, int height, int yOffsetPercentage,
+                                                      Point point)
+    {
+        int y = calculateCoordinate(height, yOffsetPercentage);
+        return createAdjustedCoordinates(startX, y, endX, y, point);
```
[the remark this answers is not in the document]
Comment (VolhaHurynovich) [reply vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/model/SwipeDirection.java]: done

File: vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/steps/TouchSteps.java (hunk 1 above)
Comment (VolhaHurynovich) [reply vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/steps/TouchSteps.java]: done

File: vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/action/TouchActions.java (hunk 2 above)
Comment (VolhaHurynovich) [reply vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/action/TouchActions.java]: done

File: vividus-plugin-mobile-app/src/test/java/org/vividus/mobileapp/steps/TouchStepsTests.java (hunk 3 above)
Comment (VolhaHurynovich) [reply vividus-plugin-mobile-app/src/test/java/org/vividus/mobileapp/steps/TouchStepsTests.java]: fixed

File: vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/steps/TouchSteps.java [hunk 6]
```diff
@@ -134,26 +135,38 @@ private Supplier<Rectangle> createSwipeArea()
         return () -> new Rectangle(new Point(0, 0), genericWebDriverManager.getSize());
     }
 
-    private void adjustVerticalPosition(WebElement element, Rectangle swipeArea, Duration swipeDuration)
+    private void adjustElementPosition(WebElement element, Rectangle swipeArea, Duration swipeDuration,
```
Comment (ikalinin1) [inline vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/steps/TouchSteps.java:138]: I propose to introduce some class which will wrap area and direction and contain the logic of swipeArea size calculation, indents calculation etc

Comment (VolhaHurynovich) [reply vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/steps/TouchSteps.java:138]: @valfirst What do you think about introducing new class?

File: vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/steps/TouchSteps.java [hunk 7]
```diff
@@ -117,9 +118,9 @@ public void swipeToElement(SwipeDirection direction, Locator locator, Duration s
             });
         }
         if (baseValidations.assertElementsNumber(String.format("The element by locator %s exists", locator), elements,
-                ComparisonRule.EQUAL_TO, 1) && (SwipeDirection.UP == direction || SwipeDirection.DOWN == direction))
+                ComparisonRule.EQUAL_TO, 1))
         {
-            adjustVerticalPosition(elements.get(0), swipeArea.get(), swipeDuration);
+            adjustElementPosition(elements.get(0), swipeArea.get(), swipeDuration, direction);
```
Comment (valfirst) [inline vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/steps/TouchSteps.java:123]:
```suggestion
            adjustElementPosition(elements.get(0), direction, swipeArea.get(), swipeDuration);
```

**Review by valfirst — CHANGES_REQUESTED** [review]

File: vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/steps/TouchSteps.java [hunk 8]
```diff
@@ -134,26 +135,38 @@ private Supplier<Rectangle> createSwipeArea()
...
             return;
         }
-
-        int topVisibilityIndent = (int) (VISIBILITY_TOP_INDENT_COEFFICIENT * swipeAreaSizeHeight);
-        if (elementTopCoordinateY < topVisibilityIndent)
+        if (elementCoordinate < topOrRightVisibilityIndent)
         {
-            touchActions.performVerticalSwipe(swipeAreaCenterY,
-                    swipeAreaCenterY + topVisibilityIndent - elementTopCoordinateY, swipeArea, swipeDuration);
+            touchActions.performSwipe(topOrRightVisibilityIndent, 2 * topOrRightVisibilityIndent - elementCoordinate,
```
[the remark this answers is not in the document]
Comment (VolhaHurynovich) [reply vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/steps/TouchSteps.java:168]: topOrRightVisibilityIndent + (topOrRightVisibilityIndent - elementCoordinate)

File: vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/action/TouchActions.java [hunk 9]
```diff
@@ -215,17 +215,26 @@ private Supplier<BufferedImage> createScreenShooter(SearchContext searchContext)
     }
 
     /**
-     * Performs vertical swipe from <b>startY</b> to <b>endY</b> with <b>swipeDuration</b>
+     * Performs vertical or horizontal swipe from <b>start</b> to <b>end</b> coordinate with <b>swipeDuration</b>
      *
-     * @param startY        start Y coordinate
-     * @param endY          end Y coordinate
+     * @param start         start Y or X coordinate
```
[the remark this answers is not in the document]
Comment (VolhaHurynovich) [reply vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/action/TouchActions.java]: done

File: vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/action/TouchActions.java [hunk 10]
```diff
@@ -215,17 +215,26 @@ private Supplier<BufferedImage> createScreenShooter(SearchContext searchContext)
...
+     * @param end           end Y or X coordinate
      * @param swipeArea     the area to execute the swipe
+     * @param direction     direction to swipe, either <b>UP</b> or <b>DOWN</b> or <b>LEFT</b> or <b>RIGHT</b>
      * @param swipeDuration swipe duration in <a href="https://en.wikipedia.org/wiki/ISO_8601">ISO 8601</a> format
      */
-    public void performVerticalSwipe(int startY, int endY, Rectangle swipeArea, Duration swipeDuration)
+    public void performSwipe(int start, int end, Rectangle swipeArea, SwipeDirection direction, Duration swipeDuration)
```
[the remark this answers is not in the document]
Comment (VolhaHurynovich) [reply vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/action/TouchActions.java]: done

File: vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/action/TouchActions.java [hunk 11]
```diff
@@ -215,17 +215,26 @@ private Supplier<BufferedImage> createScreenShooter(SearchContext searchContext)
...
     {
-        swipe(SwipeDirection.createCoordinates(startY, endY, swipeArea.getWidth(),
-                mobileApplicationConfiguration.getSwipeVerticalXPosition(), swipeArea.getPoint()), swipeDuration);
+        if (SwipeDirection.LEFT == direction || SwipeDirection.RIGHT == direction)
+        {
+            swipe(SwipeDirection.createCoordinates(start, end, swipeArea.getHeight(),
+                    mobileApplicationConfiguration.getSwipeHorizontalYPosition(), swipeArea.getPoint(), direction),
+                    swipeDuration);
```
[the remark this answers is not in the document]
Comment (VolhaHurynovich) [reply vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/action/TouchActions.java]: thank you, done

File: vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/model/SwipeDirection.java [hunk 12]
```diff
@@ -31,7 +31,7 @@ public SwipeCoordinates calculateCoordinates(Rectangle swipeArea,
         {
             int indent = swipeArea.getHeight() / VERTICAL_INDENT_COEFFICIENT;
             return createCoordinates(swipeArea.getHeight() - indent, indent, swipeArea.getWidth(),
-                    mobileApplicationConfiguration.getSwipeVerticalXPosition(), swipeArea.getPoint());
+                    mobileApplicationConfiguration.getSwipeVerticalXPosition(), swipeArea.getPoint(), UP);
```
[the remark this answers is not in the document]
Comment (VolhaHurynovich) [reply vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/model/SwipeDirection.java]: fixed

File: vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/steps/TouchSteps.java (hunk 7 above)
Comment (VolhaHurynovich) [reply vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/steps/TouchSteps.java]: done

File: vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/steps/TouchSteps.java [hunk 13]
```diff
@@ -134,26 +135,38 @@ private Supplier<Rectangle> createSwipeArea()
...
-        int swipeAreaSizeHeight = swipeArea.getHeight();
-        int swipeAreaCenterY = swipeAreaSizeHeight / 2;
-        int elementTopCoordinateY = element.getLocation().getY();
-
-        int bottomVisibilityIndent = (int) (VISIBILITY_BOTTOM_INDENT_COEFFICIENT * swipeAreaSizeHeight);
-        int visibilityY = swipeAreaSizeHeight - bottomVisibilityIndent;
-        if (elementTopCoordinateY > visibilityY)
+        int swipeAreaSize = swipeArea.getWidth();
+        int elementCoordinate = SwipeDirection.LEFT == direction
+                ? element.getLocation().getX() + element.getSize().getWidth() : element.getLocation().getX();
```
[the remark this answers is not in the document]
Comment (VolhaHurynovich) [reply vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/steps/TouchSteps.java]: done

File: vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/action/TouchActions.java [hunk 14]
```diff
@@ -215,17 +215,29 @@ private Supplier<BufferedImage> createScreenShooter(SearchContext searchContext)
...
+        if (SwipeDirection.LEFT == direction || SwipeDirection.RIGHT == direction)
+        {
+            swipe(direction.createCoordinates(startCoordinate, endCoordinate, swipeArea.getHeight(),
+                    mobileApplicationConfiguration.getSwipeHorizontalYPosition(), swipeArea.getPoint()),
+                    swipeDuration);
+            return;
+        }
+        swipe(direction.createCoordinates(startCoordinate, endCoordinate, swipeArea.getWidth(),
+                mobileApplicationConfiguration.getSwipeVerticalXPosition(), swipeArea.getPoint()),
+                swipeDuration);
```
Comment (valfirst) [inline vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/action/TouchActions.java:240]: please, fix duplication

File: vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/model/SwipeDirection.java [hunk 15]
```diff
@@ -78,11 +73,17 @@ public SwipeCoordinates calculateCoordinates(Rectangle swipeArea,
...
             MobileApplicationConfiguration mobileApplicationConfiguration);
 
-    public static SwipeCoordinates createCoordinates(int startY, int endY, int width, int xOffsetPercentage,
-            Point point)
+    public SwipeCoordinates createCoordinates(int startCoordinate, int endCoordinate, int heightOrWidth,
+            int offsetPercentage, Point point)
     {
-        int x = calculateCoordinate(width, xOffsetPercentage);
-        return createAdjustedCoordinates(x, startY, x, endY, point);
+        int calculatedCoordinate = calculateCoordinate(heightOrWidth, offsetPercentage);
```
Comment (valfirst) [inline vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/model/SwipeDirection.java:79]:
```suggestion
        int swipeAxisCoordinate = calculateCoordinate(heightOrWidth, offsetPercentage);
```

File: vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/steps/TouchSteps.java [hunk 16]
```diff
@@ -42,6 +42,7 @@
 {
     private static final float VISIBILITY_TOP_INDENT_COEFFICIENT = 0.15f;
     private static final float VISIBILITY_BOTTOM_INDENT_COEFFICIENT = 0.25f;
+    private static final float VISIBILITY_LEFT_RIGHT_INDENT_COEFFICIENT = 0.1f;
```
Comment (valfirst) [inline vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/steps/TouchSteps.java:45]:
```suggestion
    private static final float VISIBILITY_LEFT_INDENT_COEFFICIENT = 0.1f;
    private static final float VISIBILITY_RIGHT_INDENT_COEFFICIENT = 0.1f;
```

File: vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/steps/TouchSteps.java [hunk 17]
```diff
@@ -98,7 +99,7 @@ public void tapByLocator(Locator locator)
      * Swipes to element in <b>direction</b> direction with duration <b>duration</b>
      * The step takes into account current context. If you need to perform swipe on the element,
      * you need to switch the context to this element.
-     * @param direction direction to swipe, either <b>UP</b> or <b>DOWN</b>
+     * @param direction direction to swipe, either <b>UP</b> or <b>DOWN</b> or <b>LEFT</b> or <b>RIGHT</b>
```
Comment (valfirst) [inline vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/steps/TouchSteps.java:102]: please, update the docs as well

File: vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/steps/TouchSteps.java [hunk 18]
```diff
@@ -134,26 +135,41 @@ private Supplier<Rectangle> createSwipeArea()
...
+            bottomOrLeftVisibilityIndent = (int) (VISIBILITY_BOTTOM_INDENT_COEFFICIENT * swipeAreaSize);
+            topOrRightVisibilityIndent = (int) (VISIBILITY_TOP_INDENT_COEFFICIENT * swipeAreaSize);
+        }
+        else
         {
-            touchActions.performVerticalSwipe(swipeAreaCenterY,
-                    swipeAreaCenterY - (elementTopCoordinateY - visibilityY), swipeArea, swipeDuration);
+            swipeAreaSize = swipeArea.getWidth();
+            elementCoordinate = SwipeDirection.LEFT == direction
+                    ? element.getLocation().getX() + element.getSize().getWidth() : element.getLocation().getX();
```
Comment (valfirst) [inline vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/steps/TouchSteps.java:156]: I believe this logic should be completely symmetric to the logic from `if`-branch. But it's not for some reason

File: vividus-tests/src/main/resources/data/tables/system/mobile_app/locators/android.table [hunk 19]
```diff
@@ -1,19 +1,21 @@
...
+|scrollViewXpath            |//android.widget.ScrollView                                                            |
+|carouselViewXpath          |//android.widget.TextView[@text='Carousel']                                            |
+|swipeableAreaXpath         |//android.widget.HorizontalScrollView                                                  |
+|textFieldValueAttribute    |text                                                                                   |
+|scrollableMenuXpath        |//android.widget.ScrollView                                                            |
+|qrCodeContainerXpath       |//android.view.ViewGroup[@resource-id='RNE__Image__children__container']               |
+|firstItemAccessibilityId   |Item 1                                                                                 |
+|secondItemAccessibilityId  |Item 2                                                                                 |
+|secondVisibleItemXpath     |//android.widget.HorizontalScrollView/android.view.ViewGroup/android.view.ViewGroup[2] |
+|firstVisibleItemXpath      |//android.widget.HorizontalScrollView/android.view.ViewGroup/android.view.ViewGroup[1] |
```
Comment (valfirst) [inline vividus-tests/src/main/resources/data/tables/system/mobile_app/locators/android.table:21]: can we use something better than index-based xpaths?

File: vividus-tests/src/main/resources/data/tables/system/mobile_app/locators/ios.table [hunk 20]
```diff
@@ -18,3 +18,5 @@
 |qrCodeContainerXpath       |//XCUIElementTypeOther[@name='RNE__Image__children__container'] |
 |firstItemAccessibilityId   |Item 1                                                          |
 |secondItemAccessibilityId  |Item 2                                                          |
+|secondVisibleItemXpath     |(//XCUIElementTypeOther[@name="Item 2 Text 2"])[1]              |
+|firstVisibleItemXpath      |(//XCUIElementTypeOther[@name="Item 1 Text 1"])[1]              |
```
Comment (valfirst) [inline vividus-tests/src/main/resources/data/tables/system/mobile_app/locators/ios.table:22]: why is index used?

**Review by valfirst — CHANGES_REQUESTED** [review]

File: vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/action/TouchActions.java (hunk 14 above)
Comment (VolhaHurynovich) [reply vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/action/TouchActions.java:240]: In the first case - `mobileApplicationConfiguration.getSwipeHorizontalYPosition()`,
in the second - `mobileApplicationConfiguration.getSwipeVerticalXPosition()`

File: vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/model/SwipeDirection.java (hunk 15 above)
Comment (VolhaHurynovich) [reply vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/model/SwipeDirection.java]: done

File: vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/steps/TouchSteps.java (hunk 16 above)
Comment (VolhaHurynovich) [reply vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/steps/TouchSteps.java]: done

File: vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/steps/TouchSteps.java (hunk 17 above)
Comment (VolhaHurynovich) [reply vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/steps/TouchSteps.java:103]: docs have already been updated:

docs

File: vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/steps/TouchSteps.java (hunk 18 above)
Comment (VolhaHurynovich) [reply vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/steps/TouchSteps.java:158]: fixed UP-DOWN logic

File: vividus-tests/src/main/resources/data/tables/system/mobile_app/locators/android.table (hunk 19 above)
Comment (VolhaHurynovich) [reply vividus-tests/src/main/resources/data/tables/system/mobile_app/locators/android.table:21]: There is not much choice

android

File: vividus-tests/src/main/resources/data/tables/system/mobile_app/locators/ios.table (hunk 20 above)
Comment (VolhaHurynovich) [reply vividus-tests/src/main/resources/data/tables/system/mobile_app/locators/ios.table:22]: ios

File: vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/action/TouchActions.java [hunk 21]
```diff
@@ -215,17 +215,22 @@ private Supplier<BufferedImage> createScreenShooter(SearchContext searchContext)
...
+     * @param direction         direction to swipe, either <b>UP</b> or <b>DOWN</b> or <b>LEFT</b> or <b>RIGHT</b>
+     * @param swipeDuration     swipe duration in <a href="https://en.wikipedia.org/wiki/ISO_8601">ISO 8601</a> format
      */
-    public void performVerticalSwipe(int startY, int endY, Rectangle swipeArea, Duration swipeDuration)
+    public void performSwipe(SwipeDirection direction, int startCoordinate, int endCoordinate, Rectangle swipeArea,
+            Duration swipeDuration)
     {
-        swipe(SwipeDirection.createCoordinates(startY, endY, swipeArea.getWidth(),
-                mobileApplicationConfiguration.getSwipeVerticalXPosition(), swipeArea.getPoint()), swipeDuration);
+        swipe(direction.createCoordinates(startCoordinate, endCoordinate, swipeArea.getWidth(),
```
Comment (valfirst) [inline vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/action/TouchActions.java:231]: why is it always width?

Comment (VolhaHurynovich) [reply vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/action/TouchActions.java]: thank you, fixed

File: vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/steps/TouchSteps.java [hunk 22]
```diff
@@ -134,26 +136,40 @@ private Supplier<Rectangle> createSwipeArea()
         return () -> new Rectangle(new Point(0, 0), genericWebDriverManager.getSize());
     }
 
-    private void adjustVerticalPosition(WebElement element, Rectangle swipeArea, Duration swipeDuration)
+    private void adjustElementPosition(WebElement element, SwipeDirection direction, Rectangle swipeArea,
+            Duration swipeDuration)
     {
-        int swipeAreaSizeHeight = swipeArea.getHeight();
-        int swipeAreaCenterY = swipeAreaSizeHeight / 2;
```
Comment (valfirst) [inline vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/steps/TouchSteps.java:140]: where is this logic lost?

Comment (valfirst) [reply vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/steps/TouchSteps.java:140]: I would say such changes should be split into 2 commits/PRs

**Comment by valfirst** [comment]
http://chris.beams.io/posts/git-commit/

File: vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/steps/TouchSteps.java [hunk 23]
```diff
@@ -134,26 +134,28 @@ private Supplier<Rectangle> createSwipeArea()
...
+    private void adjustVerticalPosition(WebElement element, SwipeDirection direction, Rectangle swipeArea,
+            Duration swipeDuration)
     {
         int swipeAreaSizeHeight = swipeArea.getHeight();
         int swipeAreaCenterY = swipeAreaSizeHeight / 2;
-        int elementTopCoordinateY = element.getLocation().getY();
+        int elementCoordinateY = SwipeDirection.DOWN == direction
+                ? element.getLocation().getY() + element.getSize().getHeight() : element.getLocation().getY();
 
         int bottomVisibilityIndent = (int) (VISIBILITY_BOTTOM_INDENT_COEFFICIENT * swipeAreaSizeHeight);
```
Comment (valfirst) [inline vividus-plugin-mobile-app/src/main/java/org/vividus/mobileapp/steps/TouchSteps.java]: there is no sense to check bottom if direction is `DOWN`, and vice versa

**Review by valfirst — APPROVED** [review]

**Review by ikalinin1 — APPROVED** [review]

**Review by uarlouski — APPROVED** [review]

**Pull request merged**